cache.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 Adam Tauber, Andrzej Lichnerowicz
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package colly
+
+import (
+	"crypto/sha1"
+	"encoding/gob"
+	"encoding/hex"
+	"errors"
+	"net/http"
+	"os"
+	"path"
+)
+
+// Cache is an interface which handles caching Collector's responses
+// The default Cache of the Collector is the NullCache.
+// FileSystemCache keeps compatibility with non-pluggable caching in legacy
+// Collector. For this reason, one can set cache folder via `CACHE_DIR`
+// environment variable, or by passing CacheDir to NewCollector.
+// Collector's caching backend can be changed too by calling new method
+// Collector.SetCache
+type Cache interface {
+	// Init initializes the caching backend
+	Init() error
```

Comment:
Do we really need `Init()` as a part of the interface? This is not typical in Go. Can't we just assume that whateve `Cache` implementation is passed to colly, it's already initialized?

Reply:
The main reason I've decided to go with `Init` is consistency - it may not be idiomatic for Go, but it is used by `colly` all over the place... `Collector`, `httpBackend`, etc. My other reason was the way I was thinking about backwards compatibility - i.e. passing `SetCollector` via `NewCollector`, but after some thinking, I may call `NewFileSystemCache` just as well. If you'd like me to remove `Init` just let me know.

Reply:
> I may call NewFileSystemCache just as well. If you'd like me to remove Init just let me know.

Yeah, that sounds much better. Please do.